Cria nova forma de gerar/imprimir boletos para o Santander com repasse de dados

thiagopecanha opened the pull request:
## Motivação :muscle:
**Adiciona nova forma de impressão de boletos que recebe dados externos sem reaplicar no Brcobranca regras referentes ao código de barras, linha digitável, codigo do cedente e nosso numero, Altera tamanho de fonte e adiciona informações como documento e endereço completo ao campo `Beneficiário` do boleto, exigência da `Febraban`**

## Como testar
1 - Rodar os testes unitários.
2 - Rodar o teste `spec/brcobranca/santander_com_repasse_de_dados_spec.rb` colocar um `breakpoint` após a linha `119`.
3 - Em outro terminal ir até o seguinte diretório: `/tmp` e abrir o arquivo com nome foobar**SEQUÊNCIA DE CARACTERES**.pdf
4 - O boleto deve ser exibido conforme o esperado.

## Solução proposta :wrench:
Implementação de uma nova forma de impressão do boleto Santander que recebe os dados referentes a código de barras, linha digitável e nosso número de fonte externa retirando a necessidade de criarmos a lógica e regra de negócio aplicada a estes atributos do nosso lado.

## Riscos :warning:
1 - Os boletos deixarem de serem gerados.
2 - Os boletos serem gerados com as informações incorretas.

## Impactos previstos :boom:
**Informe o impacto previsto que o deploy deste PR gerará**
- Todos os boletos passarão a informar os dados de nome do beneficiário, documento e endereço completo no campo `Beneficiário` do boleto, o tamanho da fonte do texto desses dados é reduzida.
- Somente os boletos registrados através do **Santander Webservice** serão `gerados/impressos` de acordo com o repasse de regras de negócio(geradas pelo detentor do conhecimento Santander) para o código de barras, linha digitável, cedente e nosso número. 

## Observações:
- É necessário criar um débito técnico para pensarmos na construção, manutenção, atualização e refatoração dessa `gem`, temos alguns problemas relacionados a acoplamento e mal uso de herança.

thiagopecanha committed:
Adiciona novo boleto Santander

- Adiciona ao Brcobranca novo modelo de gerador de boleto Santander
que recebe  informações e as repassa para o template
-  Adiciona uma nova classe pai sem necessidade de sobrescrever metodos
para o novo boleto Santander
-  Adiciona nova classe para o Santander que efetua algumas validações
- Reformata o boleto com fontes menores para o cedente e corrige
pontos que estavam com erros

- Será necessário refatorar as classes pais para torna-las
aderentes a dois principios de design patterns sOLid

thiagopecanha committed:
Repassa os números recebidos da carteira e
conta corrente

laerte-guimaraes commented on lib/brcobranca.rb:
```diff
@@ -96,14 +96,16 @@ def self.setup
     autoload :Caixa,         'brcobranca/boleto/caixa'
     autoload :Sicoob,        'brcobranca/boleto/sicoob'
     autoload :Sicredi,       'brcobranca/boleto/sicredi'
+    autoload :SantanderComRepasseDeRegras, 'brcobranca/boleto/santander_com_repasse_de_regras'
```
O que acha de colocar `SantaderWebservice` ?

thiagopecanha replied:
Então :blush:, O **Brcobranca** deve ser agnóstico, a responsabilidade dele é receber os dados e gerar boletos eu posso repassar esses dados através de um arquivo txt, banco de dados ou etc.. então não faz sentido definirmos e limitarmos como *Webservice*.

laerte-guimaraes replied:
Então vc teria alguma outra ideia de nome?
Esse nome pra mim não faz muito sentido :thinking:

vitorkusiaki replied:
Concordo com o Laerte nessa. Estamos assumindo que o boleto pode vir de várias fontes, mas não sabemos. 
Neste caso, estamos especificamente lidando com um boleto webservice. Se houver o caso de receber de outra forma, acho válido deixar genérico.

thiagopecanha replied:
Conversamos, @laerte-guimaraes, @augustohub  e chegamos a conclusão que o melhor nome seria `SantanaderV2`.

st3llaris replied:
@thiagopecanha Faria sentido ter um nome parecido com `SantanderHandler`?

thiagopecanha replied:
Renomeamos para SantanderV2, it's done.

laerte-guimaraes commented on lib/brcobranca.rb:
```diff
@@ -96,14 +96,16 @@ def self.setup
   module Boleto
-    autoload :Base,          'brcobranca/boleto/base'
+    autoload :Base, 'brcobranca/boleto/base'
+    autoload :BaseSemOverride, 'brcobranca/boleto/base_sem_override'
```
O que acha de colocar `BaseWebservice` ? :thinking:

thiagopecanha replied:
Segue o mesmo raciocínio do comentário https://github.com/vindi/brcobranca/pull/7#discussion_r340894821 , :nerd_face:

laerte-guimaraes replied:
Então vc teria alguma outra ideia de nome?
Esse nome pra mim não faz muito sentido :thinking:

vitorkusiaki replied:
Concordo com o Laerte nessa. Estamos assumindo que o boleto pode vir de várias fontes, mas não sabemos.
Neste caso, estamos especificamente lidando com um boleto webservice. Se houver o caso de receber de outra forma, acho válido deixar genérico.

thiagopecanha replied:
Conversamos, @laerte-guimaraes, @augustohub  e chegamos a conclusão que o melhor nome seria `SantanaderV2`.

laerte-guimaraes replied:
E o `BaseV2` :smiley:

thiagopecanha replied:
Boa @laerte-guimaraes, conversando com o @victorsinged ele sugeriu que utilizemos V2 e adicionemos um comentário, continua ruim mas é melhor...

thiagopecanha replied:
Done!

laerte-guimaraes commented on lib/brcobranca/boleto/santander_com_repasse_de_regras.rb:
```diff
@@ -0,0 +1,72 @@
+      validates_length_of :numero_documento, maximum: 12,
+                                             message: 'deve ser menor ou igual a 12 dígitos.'
+      validates_length_of :seu_numero, maximum: 15, message: 'deve ser menor ou igual a 15 dígitos.'
+      validates_length_of :nosso_numero_boleto, maximum: 12,
```
Aqui seria 12 né?
https://github.com/vindi/active_merchant/pull/229/files#diff-8727d3a2cac78be5b635df4c7022d3a0R443

thiagopecanha replied:
Boa manolo :metal:

laerte-guimaraes replied:
Na verdade de acordo com o link seria 13 kk

laerte-guimaraes commented on lib/brcobranca/boleto/santander_com_repasse_de_regras.rb:
```diff
@@ -0,0 +1,72 @@
+  module Boleto
+    class SantanderComRepasseDeRegras < BaseSemOverride
+      attr_accessor :seu_numero, :codigo_cedente, :codigo_barras
+      attr_reader :linha_digitavel, :nosso_numero_boleto
```
seria possível manter o nome do parâmetro como `numero_documento` ao invés de `nosso_numero_boleto` ?

thiagopecanha replied:
Como eu te expliquei mostrando o codebase, faz sentido manter o padrão que já e utilizado atualmente no **Brcobranca**

Mas, seu comentário fez eu remover algumas solitações de implementação de `override` que não são utilizadas. :wave:

laerte-guimaraes commented on lib/brcobranca/boleto/santander_com_repasse_de_regras.rb:
```diff
@@ -0,0 +1,72 @@
+
+      validates_length_of :agencia, maximum: 4, message: 'deve ser menor ou igual a 4 dígitos.'
+      validates_length_of :convenio, maximum: 9, message: 'deve ser menor ou igual a 9 dígitos.'
+      validates_length_of :numero_documento, maximum: 12,
```
Se o comentário acima não for acatado deve ser removida essa linha certo?

thiagopecanha replied:
Tens razão, gajo :portugal:

thiagopecanha replied:
Done!

thiagopecanha committed:
Remove metodos e corrige teste

- Adiciona validação de presença para atributo carteira

- Remove métodos de override desnecessários da classe base sem override

- Adiciona valor a carteira em cenário de teste que valida a criação de objeto com valores válidos e
remove carteira de teste  em cenário que validava
carteira hardcoded quando não era passado valor nenhum.

thiagopecanha committed:
Ajustando validação e removendo atributo
desnecessário

- Altera validação do nosso número boleto para 13 e remove atributo
numero document

vitorkusiaki commented on lib/brcobranca/boleto/template/rghost.rb:
```diff
@@ -171,6 +172,8 @@ def modelo_generico_cabecalho(doc, boleto)
         def modelo_generico_rodape(doc, boleto)
           # INICIO Segunda parte do BOLETO BB
           # LOGOTIPO do BANCO
+          texto_formatado_cedente = "#{boleto.cedente} - #{boleto.documento_cedente.formata_documento}"
```
```suggestion
          formatted_assignor_text = "#{boleto.cedente} - #{boleto.documento_cedente.formata_documento}"
```
O que acha?

thiagopecanha replied:
Eu acho muito melhor da forma que você sugeriu, mas podemos quebrar o padrão e misturar `pt_BR` com `en` no codebase que está todo em `pt_BR`?

vitorkusiaki replied:
justo

thiagopecanha committed:
Renomeia nomes de classes e satisfaz comentários
PR

thiagopecanha committed:
Renomeia classes no setup do Brcobranca

thiagopecanha committed:
Adiciona comentários as classes do Santander V2

laerte-guimaraes committed:
Identa linhas
